### PR TITLE
Millard muscle curves deduplication

### PR DESCRIPTION
**Pull request by pepbos**
This PR de-duplicates the data held in `SmoothSegmentedFunction` by implementing a global cache for its data, such that when the cache is given identical curve parameters, it returns a read-only, shared copy of the curve data.

The significance of de-duplicating `SmoothSegmentedFunction` is that it's used extensively by muscle components such as `Millard2012EquilibriumMuscle`. The Millard muscle uses 5 different curves, but between individual muscle instances these curves are typically identical. Taking the `RajagopalModel`  as example:

We find the following `SmoothSegmentedFunction` curves in this model, each holding multiple `SimTK::Spline`:
- `ActiveForceLengthCurve` -> 5 splines
- `ForceVelocityInverseCurve` -> 4 splines
- `FiberForceLengthCurve` -> 2 splines
- `TendonForceLengthCurve` -> 2 splines
- `ForceVelocityCurve` -> 4 splines

With each spline ~ 200 doubles = 1.6 kB
Before deduplication the Rajagopal model has ~80 copies of each of the above curves which sums up to ~2MB. Applying deduplication reduces this to ~27kB.

### Brief summary of changes
The changes apply to `SmoothSegmentedFunction.h/cpp`:
- `SmoothSegmentedFunctionParameters` is added to collect all parameters that are used to construct a new smooth function, e.g. `_x0`, `_x1` etc...
- `SmoothSegmentedFunctionData` is added to collect all data required to evaluate the smooth function, e.g. `_arraySplineUX_`, etc. This data can be constructed from the `SmoothSegmentedFunctionParameters`.
- `SmoothSegmentedFunctionDataCache` is added to hold an `unordered_map` of smart pointers to the data, with the hashed-parameters used as key. Whenever a muscle curve is constructed, the parameters are used to obtain a pointer to already existing data, or to construct a new one.
- `SmoothSegmentedFunction` holds a shared pointer to the `SmoothSegmentedFunctionData`.

### Performance Results

Performance was evaluated against current master:

| Test Name | lhs [secs] | σ [secs] | rhs [secs] | σ [secs] | Speedup |
| -------------------------- | ---------- | -------- | ---------- | -------- | ------- |
|`ToyDropLanding_nomuscles` | 0.03 | 0.00 | 0.03 | 0.00 | 1.10 | |
|`passive_dynamic` | 1.79 | 0.00 | 1.79 | 0.00 | 1.00 |
| `RajagopalModel` | 4.66 | 0.00 | 4.35 | 0.01 | 1.07 | 
| `Arm26` | 0.11 | 0.00 | 0.10 | 0.00 | 1.02 | 
| `Gait2354` | 0.14 | 0.00 | 0.14 | 0.00 | 1.02 | 
| `ToyDropLanding` | 4.74 | 0.00 | 4.70 | 0.00 | 1.01 | 
| `passive_dynamic_noanalysis` | 1.17 | 0.00 | 1.18 | 0.00 | 1.00 |

As expected, the `RajagopalModel` can be seen to have the most significant performance boost, because it has many `Millard2012EquilibriumMuscle`s.

### CHANGELOG.md

- updated.

- - -
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3442)

### CODE REVIEW

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:57]: Consider putting separate parameters on separate lines, current formatting seems random.

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:77]: This symbol is only used in this cpp file, so put it in an anonymous namespace or `static`-qualify it

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:84]: Use bracers for each block here and put `return` statements on separate lines /w indenting, rather than putting it on the same line as another block creator (the `if`)

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:90]: Anonymous namespace or `static`

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:78]: Reference qualifier usually goes on the type, not the identifier, in C++. Putting it on the identifier is more C-style (and not what OpenSim tends to do).

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:112]: `static` or anonymous namespace. Naming convention in OpenSim is `T const& v`, not `T const &v`. `std::size_t` is more correct, but other OpenSim code tends to use `size_t`.

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:118]: Same as above

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:128]: As above

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:129]: You don't need to seed this hash with the number of rows and the number of columns: iterating over the contents and combining that with a seed should be enough (or returning the seed as-is if the matrix is empty).

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:95]: Performing the comparison out-of-order compared to the class definition is a premature optimization that makes it harder to match the list of elements here with the member data list (e.g. to see if something is missing from the equality calculation)

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:144]: Format on separate lines - the verticality of the columns being X/Y isn't as important as being able to compare this list of parameters to the class definition's list of members (e.g. to ensure nothing is left out of the hashing function).

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:161]: Why not define the implementation right here? The only reason to split the declaration/definition is in case the definition uses a symbol that's defined between the two (I can't see one?).

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:195]: Why do you need a user-defined destructor that does nothing?

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:205]: `auto` west of a `std::make_shared<T>` is permissible because C++ developers will be familiar with the value the right-hand side of the expression returns.

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:208]: Even better, just make the rhs of this expression `std::make_shared<......` to remove the local temporary

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:185]: The class's name and definition (very small) is probably a better explanation of what's going on than this comment.

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:192]: This whole class should be defined in an anonymous namespace to prevent the ctor/assign/dtor/`lookup` from leaking into the public API of the resulting compilation unit (the class isn't used outside of this file?).

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:216]: This cache strongly owns all function curves - even if no other system in OpenSim is using any. Combined with the use of a `static`-lifetime variable to hold this class, it's an unmanaged memory leak.

The `static` lifetime guarantees that the *cache* will only be destroyed at process-end, but strongly consider making the cache hold `std::weak_ptr<T>`'s rather than `std::shared_ptr<T>`'s, with the necessary `lock`ing, checking, reconstruction, etc. so that the curves are correctly freed when models are destroyed.

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:295]: `&` placement

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:372]: Break parameters onto separate lines for readability

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:376]: `_x0`, `_x1`, etc. could also be part of the deduplicated datastructure

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:379]: Ideally, `_smoothData` should be initialized via a (probably multiline) init list

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:386]: This is invalid - the implementation will segfault because this default ctor doesn't initialize `_smoothData`. You need a nullstate version of `SmoothSegmentedFunctionParameters` (i.e. you need a default ctor for `SmoothSegmentedFunctionParameters` that mimics the behavior that was observed in the old default ctor).

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (pepbos) [reply OpenSim/Common/SmoothSegmentedFunction.cpp:112]: I think you mean `const T& v` right?

Comment (pepbos) [reply OpenSim/Common/SmoothSegmentedFunction.cpp:112]: I see `using namespace std` at the top, but also `std::string` rather than `string`. Is `size_t` special in this regard?

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (pepbos) [reply OpenSim/Common/SmoothSegmentedFunction.cpp:195]: That is a very good question...

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [reply OpenSim/Common/SmoothSegmentedFunction.cpp]: Technically, yes, because `size_t` is a globally-defined data type from `<cstddef>`, but (if I understand correctly) it may not be defined, or differently defined, from `std::size_t`, which is the C++ standard's datatype for storing the size of an element in memory (for the given architecture, etc.)

Practically speaking, though, `static_assert(std::is_same_v<size_t, std::size_t>)` will pass in all codebases I've worked on up to now, and most codebases just use `size_t`.

As for `using namespace std;`, it means you can technically use `string` rather than `std::string`, but don't, because `using namespace std;` is bad practise (most codebases in the wild forbid it).

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (pepbos) [reply OpenSim/Common/SmoothSegmentedFunction.cpp]: My reasoning was, that because the matrix is hashed first, there is no seed yet. Also, it should be possible to compute the hash of just a matrix, so if it is empty I should return something. I could create a default hash in that case, but this seemed like a good alternative. But perhaps I misunderstand.

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (pepbos) [reply OpenSim/Common/SmoothSegmentedFunction.cpp:105]: I moved it to an anonumous namespace

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (pepbos) [reply OpenSim/Common/SmoothSegmentedFunction.cpp:122]: I put this in anonymous namespace as well

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:68]: Conventionally, there's a space separator between the constructor's parameter list and the initializer list (e.g. `Ctor(T arg, T arg2) : val_(arg), val2_(arg2) {}`)

(unsure why, probably initially related to your form resembling a `goto` label if you squint)

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:82]: Should this be indented?

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:130]:
```c++
auto EqualsOrBothNan = [](float a, float b) { return a == b || (std::isnan(a) && std::isnan(b); };

return
    lhs._mX == rhs._mX &&
    lhs._mY == rhs._mY &&
    EqualsOrBothNan(lhs._x0, rhs._x0) &&
    etc;
```

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:296]: Nitpick: rename to `garbageCollectExpiredData`

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:322]: Formatting: `if (expr)`

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:427]: As above (`Ctor(args) : list`)

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:563]: When using references or pointers, always keep `const-` qualification in mind. This should be:

```c++
const auto& arraySplineUX = _smoothData->arraySplineUX;
```

And the typename isn't very long, so you can make it more readable in isolation by removing `auto`:

```c++
const SimTK::Array_<SimTK::Spline>& arraySplineUX = _smoothData->_arraySplineUX;
```

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:733]: `const`

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:811]: `const`

**Review by adamkewley — COMMENTED** [review]
2nd-pass review: mostly minor stuff

File: OpenSim/Common/SmoothSegmentedFunction.cpp
Comment (adamkewley) [inline OpenSim/Common/SmoothSegmentedFunction.cpp:33]: `cmath`: `math.h` is C, `cmath` is C++ (and usually aliases `math.h`)

**Comment by pepbos** [comment]
Thanks @adamkewley for your comments!
@nickbianco, @aymanhab or @carmichaelong would you be available for reviewing this?

**Comment by nickbianco** [comment]
Hi @pepbos! I'm happy to review. Will leave some comments soon.

**Comment by pepbos** [comment]
Great! Thanks for making time!

**Review by nickbianco — APPROVED** [review]
LGTM! Seems like Adam very thoroughly reviewed this already.

Cool changes. I learned some stuff about smart pointer management.

> Reviewed 1 of 2 files at r2, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3442#-:-NVGHUCJ0SncomUhoLLc:b-f80pnr)* status: all files reviewed, 35 unresolved discussions (waiting on @adamkewley and @pepbos)

**Comment by nickbianco** [comment]
Once the CI passes, we can merge.

File: OpenSim/Common/SmoothSegmentedFunction.h
Comment (nickbianco) [inline OpenSim/Common/SmoothSegmentedFunction.h:34]: Based on the CI failure for Windows, I think this forward declaration should be `struct SmoothSegmentedFunctionData`.

**Review by nickbianco — CHANGES_REQUESTED** [review]
One small fix to get the CI to pass.

File: OpenSim/Common/SmoothSegmentedFunction.h
Comment (pepbos) [reply OpenSim/Common/SmoothSegmentedFunction.h]: Right! I think that would fix it, I pushed the change.

**Comment by pepbos** [comment]
> LGTM! Seems like Adam very thoroughly reviewed this already.
> 
> Cool changes. I learned some stuff about smart pointer management.
> 
> > Reviewed 1 of 2 files at r2, 2 of 2 files at r6, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3442#-:-NVGHUCJ0SncomUhoLLc:b-f80pnr)_ status: all files reviewed, 35 unresolved discussions (waiting on @adamkewley and @pepbos)

Great! Thanks for reviewing! I hope the CI will pass now.

**Review by nickbianco — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3442)* status: all files reviewed, 35 unresolved discussions (waiting on @adamkewley and @pepbos)

**Comment by nickbianco** [comment]
Thanks @pepbos!

**Pull request merged**
